File: Commons/XML/src/main/java/com/greentree/common/xml/parser/XMLScanner.java
package com.greentree.common.xml.parser;

import static com.greentree.common.util.iterator.IteratorUtil.filter;
import static com.greentree.common.util.iterator.IteratorUtil.iterable;
import static com.greentree.common.util.iterator.IteratorUtil.map;
import static com.greentree.common.util.iterator.IteratorUtil.min;

import java.util.Iterator;
import java.util.NoSuchElementException;

import com.greentree.common.util.cortege.Pair;

public class XMLScanner implements Iterator<XMLTocen> {

	private final String code;
	private int position;

	public XMLScanner(String code) {
		int index = -1;
		do {
			index = code.indexOf("<!--", index + 1);
			if(index == -1) break;
			final var e = code.indexOf("-->", index);
			code = code.substring(0, index) + code.substring(e + 3);
			index = e;
		}while(true);
		this.code = code.replace('\n', ' ').replace('\r', ' ').replace('\t', ' ').replace("  ", " ");
	}

	@Override
	public boolean hasNext() {
		return position < code.length();
	}

	@Override
	public XMLTocen next() {
		final var p1 = iterable(XMLTocenType.values());
		final var p2 = map(p1, tt-> {
			final var m = tt.find(position, code);
			if(m == null) return null;
			return new Pair<>(tt, m);
		});
		final var p3 = filter(p2, p->p != null);
		final var pr = min(p3, p->p.seconde.end());

		if(pr != null) {
			final var m = pr.seconde;
			position = m.end();
			final var t = new XMLTocen(m.substring(code), pr.first);
			return t;
		}
		throw new NoSuchElementException("position=" + position + " length=" + code.length() + " " + code.substring(position));
	}

}

File: Commons/XML/src/main/java/com/greentree/commons/xml/parser/token/XMLScanner.kt
package com.greentree.commons.xml.parser.token

import java.util.*

private fun removeUnUsed(text: CharSequence, builder: StringBuilder = StringBuilder()): StringBuilder {
	run {
		val index = text.indexOf("<!--")
		if(index != -1) {
			val e = text.indexOf("-->", index)
			removeUnUsed(text.substring(0, index), builder)
			removeUnUsed(text.substring(e + 3), builder)
			return builder
		}
	}
	run {
		val index = text.indexOf("<?")
		if(index != -1) {
			val e = text.indexOf("?>", index)
			removeUnUsed(text.substring(0, index), builder)
			removeUnUsed(text.substring(e + 2), builder)
			return builder
		}
	}
	builder.append(text)
	return builder
}

class XMLScanner(code: String) : Iterator<XMLTocen> {

	private var code: CharSequence =
		removeUnUsed(code).toString()

	override fun hasNext(): Boolean {
		return 0 < code.length || next.isNotEmpty()
	}

	private val next: Queue<XMLTocen> = LinkedList()

	override fun next(): XMLTocen {
		if(next.isNotEmpty())
			return next.remove()
		val matcher = xmlTocenTypes
			.mapNotNull {
				it.find(code)
			}
			.minByOrNull {
				it.end
			} ?: throw NoSuchElementException("$code")
		code = code.subSequence(matcher.end ..< code.length)
		next.addAll(matcher.token)
		return next.remove()
	}
}
